### ISSUE TITLE
LambdaContainer misbehaviour when the source contains duplicate items

### ISSUE THREAD

**Issue by vinesworth** [wasabii/OLinq#9]
A [LambdaContainer](https://github.com/wasabii/OLinq/blob/master/OLinq/LambdaContainer.cs#L25-L27) is effectively a `Dictionary`, keyed by the source elements:

``` c#
        IEnumerable<TSource> source;
        Dictionary<TSource, LambdaOperation<TResult>> lambdas =
            new Dictionary<TSource, LambdaOperation<TResult>>();
```

When the source contains duplicate elements, they get merged into a single LambdaOperation, which has some undesireable consequences:
1. if one of the duplicate items gets removed from the source, the LambdaOperation is removed immediately, which is equivalent to silently removing all of the duplicate items;
2. if the duplicated item raises OnPropertyChanged, the merged LambdaOperation will retranslate the notification only once, instead of one-per-duplicate, which will mislead the group operations like Sum, etc.

At the first glance it looks like simple reference counting would suffice, but I've discovered this issue while trying to make GroupBy notifications respect the ordering. That alone, in turn, requires knowing all of the source item indices, therefore a MultiMap-like solution would fit better.

If my vision of the situation is correct, I'll try fixing it as well :)
